Add collection position reconcilliation to card/dashboard/pulse updates

senior opened the pull request:
Cards, dashboards and pulses all have a `collection_position` that can be updated and should trigger other cards/dashboards/pulses in that collection to also be changed.

This PR adds the ability to recognize when a position changes (either via an update, or a move from one collection to another) and to fixup the collection after that update/move.

- [x] Cards
- [x] Dashboards
- [x] Pulses

iethree committed:
Add collection position reconcilliation to cards/dashboards/pulses

Anytime a new card/dashboard/pulse is created with a collection
position, or a card/dashboard/pulses has it's collection or collection
position updated, it's possible that other items in the collection
will need to have their collection position adjusted. This commit adds
a check to the PUT and POST routes of cards/dashboards/pulses to check
to see if the change impacts the position of other
cards/dashboards/pulses in the collection.

This also uses the same logic for the bulk card collection updates and
will adjust impacted cards as needed.

tlrobinson commented:
@senior Dashboards also support `collection_position`, could you support those as well?

senior commented:
Yes, this is currently WIP because dashboards and pulses aren't done. I think the code will end up being shared, but it'll take some time to get the tests going for dashboards and pulses. I'll add a TODO list so that part is more clear.

iethree committed:
Fix transient issue with pulse search test

The test query to ensure alerts aren't returned with pulse search, and
just filtering out the id of the pulse to ensure it's not in the
results. That test can fail if there happens to be a search result
with the same id as the (missing) pulse. Adding another predicate to
ensure it's only looking for pulses.

tlrobinson committed:
Merge branch 'master' of github.com:metabase/metabase into collection-position-updates

tlrobinson committed:
Update drag and drop collection_position indexes

tlrobinson committed:
More pin reordering improvements

iethree committed:
Make collection position span all items in the collection

tlrobinson committed:
Remove log

iethree committed:
Change collection position tests to use more heterogeneous collections
